### DISCUSSION TITLE
JACK/PipeWire, and overall Redesign.

### DISCUSSION THREAD

**Discussion by smoge** [smoge/Euterpea-Oxum#1] [General]
Feel free to share your thoughts.

**Comment by TristanCacqueray** [comment]
Hey @smoge, that sounds interesting! The links from the README are 404 though. I was looking for an example of the music notation.

I don't have much to share, beside perhaps, to consider Pipewire instead of Jack. I made a prototype binding [here](https://github.com/TristanCacqueray/pipewire.hs) and I'd be interested in adding support for MIDI.

**Reply by smoge** [reply]
Hi @TristanCacqueray. You have been doing excellent work; I should pay more attention to it. I will check your pipewire project now. 

Re: JACK MIDI: Two MIDI clients can communicate during the same (audio) callback process, and it's delay-free. Each MIDI event has a timestamp that says which audio sample it should line up within the buffer. That's a way to avoid timing jitter problems with MIDI in some contexts. There is nothing more efficient and precise regarding timing using MIDI than this. 

I will study your work with Pipewire; you use it for low-latency audio synthesis in Haskell. I managed to do some simple synthesis just with Haskell, but I have yet to try the kind of things Supercollider does, with Ugens building Graphs and being able to route several of those Graphs through buses, etc. To do this sort of thing, one would need to change the programming style I got used to in Haskell, and it would not be something that is not quite a typical Haskell programming experience, at least behind the scenes. It would feel like coding in C++ regarding the effort to know the most efficient ways to do things and take care of memory, etc.

One experiment I also tried was using the JACK API as my "Node Tree," as one would have in SuperCollider.  JACK, capable of parallelization and topological sorting, would handle all the Node Tree complexity for free; a Haskell program would need to create independent JACK clients. The graphs could even be compiled at runtime; why not? That was one way I managed this. I would love to hear more about your experience.

But I think you have much more to share about that. Let's be in touch. This repository here would be like a soft real-time that depends on a robust audio engine like SC or something else. 
But I'm also experimenting with other alternatives too.

**Reply by smoge** [reply]
Regarding the examples, I have kept everything that was the same as the original Euterpea and Paul Hudak's book (Haskell School of Music). I mostly fixed some build issues, switched to stack, and worked on the code without changing its functionality. I just added some tests (mostly property-based) to have something to hold on to since the original project had none. Sound testing will be necessary to redesign the core parts for more expressive and flexible music expressions and to know that everything is not falling apart. It will be interesting, too, to write tests for stateful things like events and audio signals. The original code built a lot on top of very naive ideas regarding music theory, making it only possible to do something other than tonal or 12-tone materials. And it is not even particularly better to work with that; it is simply limited. So, after the tests, I will see what design would open up this framework for different things without breaking everything else.

 [The book is here, check it out](https://www.cs.yale.edu/homes/hudak/Papers/HSoM.pdf)

Hopefully, after that, many new examples and materials will be unearthed, and new possibilities will become available.

**Reply by TristanCacqueray** [reply]
Thank you for the kind words!

I didn't meant to use Pipewire (or Jack) for audio, that sounds like a lot of work and it will be hard to match the quality of external sound synthesis engine. Perhaps Haskell can provide a nice abstraction to hide the real world IOs such as the buffering logic, like the Signal modules from the `synthesizer` or `synthesizer-llvm` seems to be doing. Though I think it's more valuable to focus on controlling an external system with MIDI or OSC.

Thinking about it some more, using Pipewire might be premature, the bindings I wrote are not battle tested, and it doesn't seem like Jack is going anywhere soon, but it might be interesting to look at what offers Pipewire, and perhaps it is a better pick long term.

I see in the book that audio synthesis is mentioned, so perhaps that is necessary to break free from western notation. Though I wonder if the same effect could be achieved with MIDI MPE, or custom CC/OSC to fine tune an external engine. Or perhaps we can generate custom wave samples live to be used by a sampler or a wavetable based synth?

It would be good to have an example to show what Euterpea is about, I'm interested in seeing how a composition looks and sounds like :)

**Reply by smoge** [reply]
@TristanCacqueray There was a good presentation today somewhat related to those things (schedulers, FFI, etc) on the Well Typed channel: https://youtu.be/IMrBTx7aYjs 

I thought it would interest you)

**Reply by TristanCacqueray** [reply]
Nice, that's very interesting. For pipewire I used inline-c, and to be honest, I don't even know which flavor of FFI it is using :) That's the same implementation as the [dear-imgui.hs](https://github.com/haskell-game/dear-imgui.hs) bindings. I forgot to share this post where I explain how it works: https://midirus.com/blog/haskell-inline-c-pipewire

I also forgot to mention this tiny library I wrote to process audio samples: https://github.com/TristanCacqueray/simple-dsp . Though I still think that's a lot of work for questionable value, and I wonder if relying on a external implementation is the best option.
